Add timeout support to regression_test.py and improve output

tothxa opened the pull request:
**Type of change**
Bugfix / New feature

**Issue(s) closed**
Fixes #1996
Fixes #5805

**New behavior**
 - Test cases are stopped and marked as failed if they don't finish within the set timeout. Default is 10 minutes.
 - Debug builds in workflow runs use 20 minutes timeout, as was found necessary.
 - Elapsed time is reported for each test case.
 - The name of the test case is flushed to stdout before the Widelands binary is started to make it easier to identify which test is running.
 - Tests are run in a deterministic order. (all `glob()`s are now `sorted()`)
 - Finding the Widelands binary in the top level of the source tree actually works.

**Possible regressions**
Test suite problems

tothxa committed:
Add timeout support to regression_test.py and improve output

tothxa commented:
In the first run, one test timed out. While that was a nice demonstration that the timeout works as intended, it's also a sign that we may need to increase the default timeout for the sake of the workflow, if we don't want frequent failures due to timeouts. Not sure though how much it should be. I think 30 minutes would be too much. 15 minutes may not help much if 10 minutes was too little. Maybe 20 minutes is tolerable? Or should the default be 10 minutes and the workflow should set it to 20 minutes with a comment?

tothxa commented:
BTW most of `regression_test.py` is just tooling to force the square peg of running a binary with arguments and check its return value into the round hole of python unit tests (and then some more to iterate over items in directories), so I tried to reimplement it in posix sh: 
[regression_test.sh.txt](https://github.com/widelands/widelands/files/10898616/regression_test.sh.txt)

Though I had to implement some basic tooling that is available in python, and almost gave up several times in the process, the end result is actually smaller in terms of characters even if it has more lines. But the main benefit is better control over its output, and (IMO) better readibility due to doing much more things natively in the implementation language, and needing much less bending over backwards to make the whole thing work.

The only missing major functionality is the ability to run additional python scripts from the test directory. Using a fixed directory for output files isn't implemented either, but that would be easy to add.

I have already tested most of the script.

matthiakl commented:
I do not know the circumstances, when these timeouts happen, but maybe it would be helpful to have some kind of time output in the log as well? Simply adding timestamps to see which tests run longer than expected.

Your script is indeed better readable (for me ;)). You might want to add a logging function to reduce the number of `echo` lines.

> The only missing major functionality is the ability to run additional python scripts from the test directory.

Is this actually used? There are no python scripts in the test directory. Maybe for additional manual tests?

tothxa commented:
> I do not know the circumstances, when these timeouts happen,

I believe the timeout in the workflow run of this PR was just caused by the many parallel runs taking up VM time from each other. But I don't know how VM provisioning works.

> but maybe it would be helpful to have some kind of time output in the log as well? Simply adding timestamps to see which tests run longer than expected.

Indeed. Unfortunately both `time` and `date` seem to be vastly incompatible between GNU and BSD, and the common posix subset is unusably limited, as always… (`time` and `timeout` would probably clash anyway, and I don't think I can run shell functions through either)

So probably timestamps are the most I could do, but that wouldn't be very helpful. Maybe test for GNU `date +%s` and BSD `date -r`, and fall back on timestamps only if neither works?

> Your script is indeed better readable (for me ;)). You might want to add a logging function to reduce the number of `echo` lines.

How do you think that would help? It would just replace all `echo`s with a call to it. Not using `\n` for empty lines is just my preference for readibility, though it seems to be shared by other shell scripts we have.

tothxa committed:
report elapsed time for each Widelands run

tothxa commented:
Sorry, I forgot that the python script is the real one. :) (for now?) Added timing info to it.

Maybe this should be taken over by somebody who actually knows python. ;) While hunting for documentation for the last commit, I've found out that `subprocess` actually supports timeout, but as far as I could tell, not the way we use it: at least `run()` and `communicate()` do, but we read stdout line by line instead. `readline()` doesn't seem to have a simple way to do timeouts, but even if it did, that wouldn't be sufficient for this use case, because there may be problems where the test script fails to finish, but the game does not hang, so it keeps outputting due to `--verbose`.

[edit:] Oops, I've just realised that I could test the elapsed time in the `while` loop. That, with a timeout on `readline()`, could work.

matthiakl reviewed (commented):
> How do you think that would help? It would just replace all echos with a call to it. Not using \n for empty lines is just my preference for readibility, though it seems to be shared by other shell scripts we have.

Ok, I understand your point. It just seemed a bit verbose to me with `echo`ing empty lines frequently.

tothxa committed:
use native timeout

tothxa commented:
Timeout support was added in python 3.3. Should I add compatibility checks for python 2.x (just disable timeout and timing support), or may I remove the other compatibility block from the beginning? (on my Debian oldstable system /usr/bin/python is still python2)

matthiakl commented:
If disabling it for Python 2 does not introduce too much special handling I would prefer this way.

tothxa committed:
compatibility checks for python < 3.3

tothxa committed:
improve search for binary

tothxa committed:
Merge branch 'master' into test-timeout

tothxa commented:
I've analysed the logs of the last 2 runs a bit. Some tests in both runs pushed the 10min limit with >8min elapsed time, but the vast majority doesn't even come close to it. It turns out, that the all tribes tests on debug builds always go above 5min, all the rest are <=2min.

```
$ rgrep -oh 'Elapsed time: 0:0.' log? | sort | uniq -c
   2412 Elapsed time: 0:00
    369 Elapsed time: 0:01
      1 Elapsed time: 0:02
     26 Elapsed time: 0:05
     19 Elapsed time: 0:06
     11 Elapsed time: 0:07
      7 Elapsed time: 0:08
      1 Elapsed time: 0:09
$ rgrep -B2 'Elapsed time: 0:0[2-9]:..' log? | grep -o 'test/.*' | sort | uniq -c
     17 test/maps/all_tribes.wmf/scripting/test_atlanteans.lua
     16 test/maps/all_tribes.wmf/scripting/test_barbarians.lua
     16 test/maps/all_tribes.wmf/scripting/test_empire.lua
     16 test/maps/all_tribes.wmf/scripting/test_frisians.lua
$ rgrep -B2 'Elapsed time: 0:0[7-9]:..' log? | grep -o 'test/.*' | sort | uniq -c
      5 test/maps/all_tribes.wmf/scripting/test_atlanteans.lua
     11 test/maps/all_tribes.wmf/scripting/test_barbarians.lua
      2 test/maps/all_tribes.wmf/scripting/test_empire.lua
      1 test/maps/all_tribes.wmf/scripting/test_frisians.lua
$ rgrep -c 'Elapsed time: 0:0[7-9]:..' log? | grep '[1-9]$' | sort
log1/testsuite (clang-7, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log1/testsuite (clang-8, ubuntu-20.04, Debug)/7_Testsuite.txt:1
log1/testsuite (Debug, ubuntu-22.04, clang-13)/7_Testsuite.txt:3
log1/testsuite (Debug, ubuntu-22.04, g++-11)/7_Testsuite.txt:1
log1/testsuite (Debug, ubuntu-22.04, g++-12)/7_Testsuite.txt:1
log1/testsuite (g++-8, ubuntu-20.04, Debug)/7_Testsuite.txt:1
log2/testsuite (clang-7, ubuntu-20.04, Debug)/7_Testsuite.txt:1
log2/testsuite (clang-8, ubuntu-20.04, Debug)/7_Testsuite.txt:1
log2/testsuite (Debug, ubuntu-22.04, clang-14)/7_Testsuite.txt:1
log2/testsuite (Debug, ubuntu-22.04, g++-11)/7_Testsuite.txt:2
log2/testsuite (g++-7, ubuntu-20.04, Debug)/7_Testsuite.txt:2
log2/testsuite (g++-8, ubuntu-20.04, Debug)/7_Testsuite.txt:1
$ rgrep -c 'Elapsed time: 0:0[5-9]:..' log? | grep '[1-9]$' | sort
log1/testsuite (clang-7, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log1/testsuite (clang-8, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log1/testsuite (Debug, ubuntu-22.04, clang-13)/7_Testsuite.txt:4
log1/testsuite (Debug, ubuntu-22.04, clang-14)/7_Testsuite.txt:4
log1/testsuite (Debug, ubuntu-22.04, g++-11)/7_Testsuite.txt:4
log1/testsuite (Debug, ubuntu-22.04, g++-12)/7_Testsuite.txt:4
log1/testsuite (g++-7, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log1/testsuite (g++-8, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log2/testsuite (clang-7, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log2/testsuite (clang-8, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log2/testsuite (Debug, ubuntu-22.04, clang-13)/7_Testsuite.txt:4
log2/testsuite (Debug, ubuntu-22.04, clang-14)/7_Testsuite.txt:4
log2/testsuite (Debug, ubuntu-22.04, g++-11)/7_Testsuite.txt:4
log2/testsuite (Debug, ubuntu-22.04, g++-12)/7_Testsuite.txt:4
log2/testsuite (g++-7, ubuntu-20.04, Debug)/7_Testsuite.txt:4
log2/testsuite (g++-8, ubuntu-20.04, Debug)/7_Testsuite.txt:4
```

tothxa committed:
increase timeout for debug builds in the workflow

tothxa commented:
It worked nicely. This time there were 3 parallel builds running, and 1 test went over 10 minutes, but not too much:

```
testsuite (Debug, ubuntu-22.04, clang-14)/7_Testsuite.txt:2023-03-08T02:12:32.4418024Z     Elapsed time: 0:10:31.438279
```

So I think 20 minutes for debug builds will work with a good safety margin, and 10 minutes is more than enough for release builds.

tothxa commented:
Regarding the line-by-line flushing that we removed to allow native timeouts: I think that makes sense for tests that may crash the whole testing rig, to get logging up to a point as close to the crash as possible. But that's not the case here, so I don't think we introduced any new risk of missing some of the logs.

matthiakl reviewed (commented):
> Regarding the line-by-line flushing that we removed to allow native timeouts: I think that makes sense for tests that may crash the whole testing rig, to get logging up to a point as close to the crash as possible. 

AFAIK, python flushes the streams on newline characters, so we should already get near the last complete line.

One more comment inline, the rest looks fine to me.

tothxa committed:
Apply suggestion from code review

Co-authored-by: matthiakl <t-m.42@mailpost.spdns.org>
